compiler/rustc_trait_selection/src/traits/select/confirmation.rs

```diff
@@ -314,45 +314,36 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
             let self_ty = self.infcx.shallow_resolve(obligation.predicate.self_ty());
             let cause = obligation.derived_cause(BuiltinDerivedObligation);
 
-            let mut obligations =
+            let (param_env, nested) =
                 if let ty::GeneratorWitness(interior) = self_ty.skip_binder().kind() {
-                    let tcx = self.tcx();
-                    let ty::GeneratorWitnessInner { tys: nested, structural_predicates } =
-                        self.infcx().replace_bound_vars_with_placeholders(*interior);
-
-                    let param_env = obligation.param_env;
-                    // Augment the param env with the projections that we know hold structurally
-                    // in the generator interiors' types.
-                    let augmented_param_env =
-                        ty::ParamEnv::new(
-                            tcx.mk_predicates(param_env.caller_bounds().iter().chain(
-                                structural_predicates.iter().map(|predicate| {
-                                    ty::Binder::dummy(predicate).to_predicate(tcx)
-                                }),
-                            )),
-                            param_env.reveal(),
-                            param_env.constness(),
+                    self.infcx().commit_unconditionally(|_| {
+                        let ty::GeneratorWitnessInner { tys: nested, predicates } =
+                            self.infcx().replace_bound_vars_with_placeholders(*interior);
+                        // FIXME(compiler-errors): Not sure if we should augment the param_env,
+                        // or just make a new param_env from these predicates...
+                        // Some tests fail if we do the former, but doing the latter doesn't
+                        // sound like it works 100% either...
+                        let new_param_env = ty::ParamEnv::new(
```

Comment:
This logic of being able to handle implication is one of the things that I wanted to add to the compiler via chalk, fwiw, but it may be possible to do it locally this way.